tests/cleanup_and_retention_test.go

```diff
@@ -1,191 +1,166 @@
-		rdb := getEmptyDatabase(t)
-		db, err := sqlx.Connect(rdb.Driver(), rdb.DSN())
-		require.NoError(t, err, "connecting to database")
-		t.Cleanup(func() { _ = db.Close() })
-
-		return icingadb.NewDb(db, logging.NewLogger(it.Logger(t).Sugar(), time.Second), &icingadb.Options{})
-	}
-	start := time.Date(2009, time.November, 10, 23, 0, 0, 0, time.UTC)
-	startMilli := start.UnixMilli()
-
-	scenarios := []struct {
-		limit        int64
-		rowsToDelete int64
-		rowsToSpare  int64
-	}{
-		{10, 9, 100},
-		{10, 10, 100},
-		{10, 100, 100},
-		{10, 1000, 100},
-		{100, 100, 100},
-		{1000, 100000, 100},
+	rdb := getDatabase(t)
+	db, err := sqlx.Open(rdb.Driver(), rdb.DSN())
+	require.NoError(t, err, "connecting to SQL database shouldn't fail")
+	t.Cleanup(func() { _ = db.Close() })
+
+	var schema string
+	switch db.DriverName() {
+	case "mysql":
+		schema = `SCHEMA()`
+	case "postgres":
+		schema = `CURRENT_SCHEMA()`
 	}
 
-	for i := range scenarios {
-		scenario := scenarios[i]
-
-		t.Run(fmt.Sprintf("%d rows with limit %d", scenario.rowsToDelete, scenario.limit), func(t *testing.T) {
-			t.Parallel()
-
-			table := fmt.Sprintf("retention_test_limit_%d_delete_%d", scenario.limit, scenario.rowsToDelete)
-
-			db := database(t)
-
-			_, err := db.Exec(fmt.Sprintf(`CREATE TEMPORARY TABLE %s (id int NOT NULL, event_time bigint NOT NULL, PRIMARY KEY (id))`, table))
-			require.NoError(t, err)
+	reten := retention{
+		Days: 7,
+		Options: map[string]int{
+			"acknowledgement": 0, // No retention
+			"comment":         1,
+			"downtime":        2,
+			// notification and state default to 7
+		},
+	}
 
-			_, err = db.Exec(fmt.Sprintf(`CREATE INDEX idx_%[1]s ON %[1]s (event_time)`, table))
+	rowsToDelete := 10000
+	rowsToSpare := 1000
+
+	for category, stmt := range retentionStatements {
+		rows, err := db.Query(db.Rebind(fmt.Sprintf(`
+SELECT column_name
+FROM information_schema.columns
+WHERE table_schema = %s AND table_name = ? AND column_name NOT IN (?, ?) AND is_nullable = ?`,
+			schema)),
+			stmt.Table, stmt.PK, stmt.Column, "NO")
+		if err != nil {
 			require.NoError(t, err)
-
-			// One row per millisecond.
-			total := scenario.rowsToDelete + scenario.rowsToSpare
-			type row struct {
-				Id   int64
-				Time int64
+		}
+		for rows.Next() {
+			var col string
+			if err := rows.Scan(&col); err != nil {
+				require.NoError(t, err)
 			}
-			values := make([]row, 0, MaxPlaceholders)
-			for j := int64(0); j < total; j++ {
-				values = append(values, row{j, startMilli + j})
-
-				if len(values) == MaxPlaceholders || j == total-1 {
-					_, err := db.NamedExec(fmt.Sprintf(`INSERT INTO %s (id, event_time) VALUES (:id, :time)`, table), values)
-					require.NoError(t, err)
-					values = values[:0]
-				}
+			if _, err := db.Exec(fmt.Sprintf(`ALTER TABLE %s DROP COLUMN %s`, stmt.Table, col)); err != nil {
+				require.NoError(t, err)
 			}
+		}
+		if err := rows.Err(); err != nil {
+			require.NoError(t, err)
+		}
```

Comment:
I'd go for `Select()` from sqlx here which can directly put the result in a `[]string` and only needs error handling once.

Reply:
Done.